TestProjects/ShaderGraph/Assets/CommonAssets/Editor/ShaderGraphUITestHelpers.cs

```diff
@@ -0,0 +1,124 @@
+                null,
+                new Type[] {typeof(Event), typeof(EventType)},
+                null);
+
+            return (EventBase)createEventMethodInfo?.Invoke(null,  new object[]{evt, evt.rawType});
+        }
+
+        public static EventBase MakeEvent(EventType type)
+        {
+            var evt = new Event() { type = type };
+            return CreateEvent(evt);
+        }
+
+        public static EventBase MakeEvent(EventType type, Vector2 position)
+        {
+            var evt = new Event() { type = type, mousePosition = position };
+            return CreateEvent(evt);
+        }
+
+        public static EventBase MakeKeyEvent(EventType type, KeyCode code, char character = '\0', EventModifiers modifiers = EventModifiers.None)
+        {
+            var evt = new Event() { type = type, keyCode = code, character = character, modifiers = modifiers};
+            return CreateEvent(evt);
+        }
+
+        public static EventBase MakeMouseMoveEvent(Vector2 mousePosition, MouseButton button = MouseButton.LeftMouse, EventModifiers modifiers = EventModifiers.None, int clickCount = 1)
+        {
+            var evt = new Event() { type = EventType.MouseMove, mousePosition = mousePosition, button = (int)button, modifiers = modifiers, clickCount = clickCount};
+            return CreateEvent(evt);
+        }
+
+        public static EventBase MakeMouseEvent(EventType type, Vector2 position, MouseButton button = MouseButton.LeftMouse, EventModifiers modifiers = EventModifiers.None, int clickCount = 1)
+        {
+            var evt = new Event() { type = type, mousePosition = position, button = (int)button, modifiers = modifiers, clickCount = clickCount};
+            return CreateEvent(evt);
+        }
+
+        public static EventBase MakeScrollWheelEvent(Vector2 delta, Vector2 position)
+        {
+            var evt = new Event
+            {
+                type = EventType.ScrollWheel,
+                delta = delta,
+                mousePosition = position
+            };
+
+            return CreateEvent(evt);
+        }
+
+        public static EventBase MakeCommandEvent(EventType type, string command)
+        {
+            var evt = new Event() { type = type, commandName = command };
+            return CreateEvent(evt);
+        }
+
+        public static void TestAllElements(VisualElement container, Action<VisualElement> test)
+        {
+            test(container);
+            for (int i = 0; i < container.hierarchy.childCount; ++i)
+            {
+                var element = container.hierarchy[i];
+                TestAllElements(element, test);
+            }
+        }
+
+        public static void SendMouseEventToVisualElement(
+            VisualElement elementToNotify,
+            EventType eventType,
+            MouseButton mouseButton = MouseButton.LeftMouse,
+            Vector2 eventPositionOffset = new Vector2())
```

Comment:
Nitpicky but to avoid unnecessary garbage, you can use Vector2.Zero as the default param value.

Reply:
Sounds good!

Reply:
Something interesting was that I was unable to use Vector2.zero as a default parameter here, due to the compiler error:
"Default parameter must be a compile time constant". I was unaware that was a language requirement.

Replaced it with the `default` keyword which still instantiates a new instance, but the other option of explicitly stating `new Vector2(0.0f, 0.0f)` failed with the same compiler error as above.